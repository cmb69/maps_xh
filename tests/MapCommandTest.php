<?php

namespace Maps;

use ApprovalTests\Approvals;
use Maps\Model\Map;
use org\bovigo\vfs\vfsStream;
use PHPUnit\Framework\TestCase;
use Plib\DocumentStore2 as DocumentStore;
use Plib\FakeRequest;
use Plib\View;

class MapCommandTest extends TestCase
{
    private array $conf;
    private DocumentStore $store;
    private View $view;

    public function setUp(): void
    {
        vfsStream::setup("root");
        $this->conf = XH_includeVar("./config/config.php", "plugin_cf")["maps"];
        $this->store = new DocumentStore(vfsStream::url("root/"));
        $map = Map::create("london", $this->store);
        $map->setCoordinates(-0.09, 51.505);
        $map->setZoom(13, 19);
        $map->addMarker(51.505, -0.09, "some info", true);
        $this->store->commit();
        $this->view = new View("./views/", XH_includeVar("./languages/en.php", "plugin_tx")["maps"]);
    }

    private function sut(): MapCommand
    {
        return new MapCommand("../", $this->conf, $this->store, $this->view);
    }

    public function testShowsMap(): void
    {
        $request = new FakeRequest();
        $response = $this->sut()("london", $request);
        Approvals::verifyHtml($response->output());
    }

    public function testReportsNonExistingMap(): void
    {
        $request = new FakeRequest();
        $response = $this->sut()("does-not-exist", $request);
        $this->assertStringContainsString("There is no “does-not-exist” map!", $response->output());
    }

    public function testDoesNotShowPrivacyFormIfConfigured(): void
    {
        $this->conf["tile_privacy"] = "";
        $request = new FakeRequest();
        $response = $this->sut()("london", $request);
        $this->assertStringNotContainsString("<form", $response->output());
    }

    public function testAgreementSetsCookieAndRedirects(): void
    {
        $request = new FakeRequest(["post" => ["maps_agree" => "1"]]);
        $response = $this->sut()("london", $request);
        $this->assertEquals(["maps_agreed", "1", 0], $response->cookie());
        $this->assertSame("http://example.com/", $response->location());
    }
}
