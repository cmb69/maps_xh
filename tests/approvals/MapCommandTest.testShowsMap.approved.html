
<script type="module" src="/maps.js?&amp;v=1.0-dev"></script>
<figure class="maps_map" data-maps-conf='{"tileUrl":"https://tile.openstreetmap.org/{z}/{x}/{y}.png","tileAttribution":"&copy; <a href=\"http://www.openstreetmap.org/copyright\">OpenStreetMap</a>","loadTiles":false,"latitude":-0.09,"longitude":51.505,"zoom":13,"maxZoom":19,"markers":[[51.505,-0.09,"some info",true]]}'>
  <figcaption>Map of London</figcaption>
  <div class="maps_map" style="width: 100%; aspect-ratio: 1/1"></div>
  <form method="post">
    <p>To display the map, tile images need to be fetched from a third-party server. Do you agree?</p>
    <button name="maps_agree" value="1">I agree!</button>
  </form>
</figure>
